TriforkSwiftExtensions/TriforkSwiftExtensions/Date/DateExtension.swift

```diff
@@ -105,9 +106,73 @@ public extension Date {
         }
         return date
     }
-
+    
+    var startOfYear: Date {
+        alignTo(resolution: .year)
+    }
+    
+    var startOfMonth: Date {
+        alignTo(resolution: .month)
+    }
+    
+    var startOfDay: Date {
+        alignTo(resolution: .day)
+    }
+    
+    var startOfHour: Date {
+        alignTo(resolution: .hour)
+    }
+    
+    var startOfMinute: Date {
+        alignTo(resolution: .minute)
+    }
+    
+    var startOfSecond: Date {
+        alignTo(resolution: .second)
+    }
+    
     func date(byAdding component: Calendar.Component, value: Int) -> Date {
         return Calendar.current.date(byAdding: component, value: value, to: self) ?? self
     }
+    
+    func date(bySubtracting component: Calendar.Component, value: Int) -> Date {
+        Calendar.current.date(byAdding: component, value: -value, to: self) ?? self
+    }
+    
+    func alignTo(resolution: Calendar.Component) -> Date {
```

Comment:
same here (wrt to the current calendar as a default paramter)
```swift
func alignTo(
        resolution: Calendar.Component,
        withCalendar: Calendar = Calendar.current
    ) -> Date {
        var dateComponents = withCalendar.dateComponents(
            Set([.year, .month, .day, .hour, .minute, .second, .nanosecond, .timeZone]),
            from: self)
        switch resolution {
        case .year:
            dateComponents.setValue(0, for: .month)
            dateComponents.setValue(0, for: .day)
            dateComponents.setValue(0, for: .hour)
            dateComponents.setValue(0, for: .minute)
            dateComponents.setValue(0, for: .second)
            dateComponents.setValue(0, for: .nanosecond)
        case .month:
            dateComponents.setValue(0, for: .day)
            dateComponents.setValue(0, for: .hour)
            dateComponents.setValue(0, for: .minute)
            dateComponents.setValue(0, for: .second)
            dateComponents.setValue(0, for: .nanosecond)
        case .day:
            dateComponents.setValue(0, for: .hour)
            dateComponents.setValue(0, for: .minute)
            dateComponents.setValue(0, for: .second)
            dateComponents.setValue(0, for: .nanosecond)
        case .hour:
            dateComponents.setValue(0, for: .minute)
            dateComponents.setValue(0, for: .second)
            dateComponents.setValue(0, for: .nanosecond)
        case .minute:
            dateComponents.setValue(0, for: .second)
            dateComponents.setValue(0, for: .nanosecond)
        case .second:
            dateComponents.setValue(0, for: .nanosecond)
        default:
            break
        }
        return NSCalendar.current.date(from: dateComponents) ?? self
    }
```
another note is that if this switch was split into seperate functions, the code would naturally happen without duplication (eg, alignToYear would set the month to 0, then call allignToMonth)